### PR TITLE
Speedup Nearest Projection Mapping

### PR DESCRIPTION
**Pull request by fsimonis**
This PR is a first attempt to improve the quadratic computational complexity of generating the nearest projection mapping (#99 #132).

The new approach indexes all the primitives into a cached `boost::geometry::index::rtree` and then queries for the nearest neighbour in the rtree. This _should_ result in a `n(log(n))` complexity for uncached and `log(n)` for cached meshes.

# Changes to the Interface
* `mesh::Triangle` and `mesh::Quad` now offer _const_ iteration over the coordinates of their vertices. (I could not iterate over Vertex as it is not copyable and geometry does not support reference point types as they need to be default constructable.)
* `query::InterpolationElement` now holds a pointer to `const mesh::Vertex`. (Improved const ~sanity~ correctness in various places.)
* `mesh::rtree` now also manages a cache of the new `PrimitiveRTree`

# Internal Changes
* I refactored out the barycentre calculation from the `FindClosest...` modules. I moved it to the `math::barycentre` namespace. The interface is now purely Eigen3. I also implemented the necessary code in `FindClostestQuad.cpp` so that only the math needs to be implemented for full support. (i.e. #153 boils down to implementing the function `math::barycentre::calcBarycentricCoordsForQuad`)
* I refactored the generation of interpolation into `InterpolationElements generateInterpolationElements( const mesh::Vertex &location, mesh::PRIMITIVE &element);`
* `InterpolationElements` inside FindClosest-Modules are now all `std::vector<InterpolationElement>`
* Indexing all primitives (Vertices, Edges, Triangles, and Quads) of a mesh using the `rtree` was a bit tricky. The `rtree` sadly does not directly support variant `Indexables`, so I had to generate bounding boxes for each primitive and index them with an extended index `PrimitiveIndex`. I also had to adapt the primitives to the boost.geometry.ring concept.

# TODO
* [x] Test the new iteration interface of Triangle and Quad (#161)
* [x] Test math::barycentre (#179)
* [x] Test the new rtree (#203)
* [x] Rename `PtrRTree` to `PtrVertexRTree`
* [ ] Verify the improved computational complexity
* [ ] Observe the memory impact of saving an AABB for each primitive

### CODE REVIEW

**Comment by fsimonis** [comment]
clang-format sneaked in quite a few formatting changes. I removed them to reduce the noise in the diff.

**Comment by floli** [comment]
Do you think it would be possible to split up these pull request in smaller chunks? Like the range adapter stuff / the tree adaption and finally the NP speed up?
That would be much easier to review and we are able to get it to develop quicker.

File: src/math/barycentre.hpp [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+  /// A vector of the n coefficients for n vertices
+  Eigen::VectorXd barycentricCoords;
+  /// The projected location vertex
+  Eigen::VectorXd projected;
+};
+
+/** Takes the corner vertices of an edge and its norm.
+ *  It then calculates the projection of a location vector and generates the barycentric coordinates for the corner points.
+ *
+ *  \param edgeA point A of the edge AB
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/math/barycentre.hpp]: :+1:

**Comment by fsimonis** [comment]
The iterator interface is up in a separate PR #161.
It is tested and the concept requirements are checked at compile-time using `BOOST_CONCEPT_ASSERT`.

**Comment by fsimonis** [comment]
I am isolation the math refactoring from this PR. I'll publish the PR mid next week.
That should significantly reduce the size of this PR.

**Comment by fsimonis** [comment]
This PR is now based on #203 and in a polished state.

Once #203 was reviewed and is merged into develop, this PR is ready for review.

**Comment by fsimonis** [comment]
Merged develop and renamed the remaining rtree types.
This PR is ready for final review.

**Pull request merged**